File: barrel_models.py
"""Machine-isolated engineering layouts, optional lengths and compatible caps.
Custom dimensions are proposed engineering values, never manufacturing approvals.
"""
import re
MAX_MODULES=60
PORT_ROLES=('feed','natural','vacuum','side','side_vacuum')
CAP_NAMES={'open':'工作开口','closed':'全封闭堵头','inject':'中心注液堵头'}

def catalog(machine,spec):
    prefix='Zylinder ' if machine=='50' else 'Zyl-'
    items=[('barrel',prefix+'0','barrel',spec['section_length']),('feed',prefix+'E','feed',spec['section_length']),('natural',prefix+'1','natural',spec['section_length']),('vacuum',prefix+'1','vacuum',spec['section_length']),('side','Zylinder S (LSB047)' if machine=='50' else 'Zyl-S (LSB 56)','side',spec['section_length']),('side_vacuum','Side vacuum','side_vacuum',spec['section_length']),('flange','Zwischenflansch','flange',spec['flange_length'])]
    if spec['heat_barrier']>0:items.append(('heat','Wärmesperre','heat',spec['heat_barrier']))
    return [dict(key=f'{machine}:{key}',machine=machine,name=name,role=role,length=length,nominal_diameter=spec['diameter'],bore_diameter=None,caps=(['open','closed','inject'] if role in ('natural','vacuum') else ['open','closed'] if role in PORT_ROLES else [])) for key,name,role,length in items]

def position(index):
    out='';index+=1
    while index:index,r=divmod(index-1,26);out=chr(65+r)+out
    return out

def normalize(machine,spec,layout):
    if not isinstance(layout,dict) or layout.get('version')!=1 or layout.get('machine')!=machine:raise ValueError('机筒配置版本或所属机型不符，50/60机模块不能混用')
    modules=layout.get('modules')
    if not isinstance(modules,list) or len(modules)>MAX_MODULES:raise ValueError(f'机筒模块最多{MAX_MODULES}个')
    known={x['key']:x for x in catalog(machine,spec)};ids=set();result=[]
    for m in modules:
        if not isinstance(m,dict) or m.get('key') not in known:raise ValueError('机筒模块不属于当前机型或未核对型号')
        item=known[m['key']];uid=m.get('uid');opened=m.get('open',True)
        if not isinstance(uid,str) or not re.fullmatch(r'[A-Za-z0-9_-]{1,80}',uid) or uid in ids:raise ValueError('机筒模块标识无效或重复')
        if not isinstance(opened,bool):raise ValueError('排气口状态必须为布尔值')
        ids.add(uid);cap=m.get('cap','open' if opened else 'closed')
        if item['caps']:
            if cap not in item['caps']:raise ValueError('堵头与当前机筒开口不兼容')
        else:
            if m.get('cap','open')!='open':raise ValueError('该模块没有可安装堵头的开口')
            cap='open'
        row=dict(uid=uid,key=m['key'],open=cap!='closed',cap=cap)
        if 'length' in m:
            length=m['length']
            if type(length)!=int:raise ValueError('节长必须为整数毫米')
            if item['role'] in ('heat','flange'):
                if length!=item['length']:raise ValueError('隔热片及法兰长度按本机标准保留')
            elif not (max(10,({'50':72,'60':80}[machine] if item['caps'] else 10))<=length<=spec['section_length']*4):raise ValueError('节长超出设计范围或小于开口投影宽度')
            row['length']=length
        result.append(row)
    out=dict(version=1,machine=machine,modules=result)
    if 'screw_length' in layout:
        n=layout['screw_length']
        if type(n)!=int or not 100<=n<=MAX_MODULES*spec['section_length']*4+1000:raise ValueError('方案螺杆目标长度无效')
        out['screw_length']=n
    return out

def rows(machine,spec,ports):
    ports=ports or {};layout=ports.get('barrel_layout');result=[];mm=0
    if layout is None:
        for source in spec['barrel_configuration']:
            r=dict(source);r.update(start=mm,length=r['mm']-mm,machine=machine,nominal_diameter=spec['diameter']);mm=r['mm'];r['open']=ports.get(r.get('port_key'),True);r['cap']='open' if r['open'] else 'closed';result.append(r)
    else:
        layout=normalize(machine,spec,layout);known={x['key']:x for x in catalog(machine,spec)}
        for i,m in enumerate(layout['modules']):
            r={**known[m['key']],**m,'pos':position(i),'start':mm};mm+=r['length'];r['mm']=mm;r['port_key']='module-'+r['uid'] if r['role'] in PORT_ROLES else '';result.append(r)
    for r in result:
        role=r['role'];cap=r['cap']
        r['annotation']='INJECT' if cap=='inject' else 'CLOSED' if cap=='closed' else {'feed':'FEED','natural':'ATM','side':'SIDE FEED','vacuum':'VAC','side_vacuum':'SIDE VAC'}.get(role,f"{r['length']} mm" if role=='heat' else '')
        r['accessory']=('侧喂料堵头' if role=='side' else '侧真空堵头' if role=='side_vacuum' else '顶部全封闭堵头') if cap=='closed' else '中心注液堵头' if cap=='inject' else '无堵头'
    return result

def target_length(machine,spec,ports):
    layout=(ports or {}).get('barrel_layout')
    return normalize(machine,spec,layout).get('screw_length',spec['element_length']) if layout is not None else spec['element_length']

def summary(machine,spec,ports):
    rs=rows(machine,spec,ports);process=[r for r in rs if r['role'] not in ('heat','flange')];active=sum(r['length'] for r in process)
    return dict(sections=len(process),active_length=active,process_d=round(active/spec['section_length']*4,3),barrel_length=sum(r['length'] for r in rs),suggested_screw_length=spec['element_length']+sum(r['length'] for r in rs)-spec['barrel_length'])

def warnings(machine,spec,ports):
    if (ports or {}).get('barrel_layout') is None:return []
    rs=rows(machine,spec,ports);length=sum(r['length'] for r in rs)
    notes=['自定义机筒配置：需复核筒体内径、间隙、轴长及开口位置，未核对前不可直接用于生产']
    if not rs:notes.append('机筒配置为空')
    if length!=spec['barrel_length']:notes.append(f"机筒总长 {length} mm，与标准 {spec['barrel_length']} mm 不符；标准螺杆轴长不自动延长")
    if sum(r['role']=='feed' and r['open'] for r in rs)!=1:notes.append('主喂料模块数量不为1')
    if sum(r['role']=='flange' for r in rs)!=1:notes.append('连接法兰数量不为1')
    if target_length(machine,spec,ports)!=spec['element_length']:notes.append('已指定方案螺杆目标长度：实际轴长、强度及驱动适配需工程复核')
    return notes

File: static/releases/v1.28/barrel-models.js
/* Machine-isolated engineering proposals; keep Python/JS validation identical. */
(function(root,factory){const api=factory();if(typeof module==='object'&&module.exports)module.exports=api;else root.BarrelModels=api;})(globalThis,function(){
 'use strict';
 const MAX_MODULES=60,PORT_ROLES=['feed','natural','vacuum','side','side_vacuum'];
 function catalog(spec){return spec.barrel_catalog||[]}
 function position(i){let out='';for(i++;i;i=Math.floor((i-1)/26))out=String.fromCharCode(65+(i-1)%26)+out;return out}
 function normalize(machine,spec,layout){
  if(!layout||layout.version!==1||layout.machine!==machine)throw Error('机筒配置版本或所属机型不符，50/60机模块不能混用');
  if(!Array.isArray(layout.modules)||layout.modules.length>MAX_MODULES)throw Error(`机筒模块最多${MAX_MODULES}个`);
  const known=new Map(catalog(spec).map(x=>[x.key,x])),ids=new Set();
  const result={version:1,machine,modules:layout.modules.map(m=>{
   if(!m||!known.has(m.key))throw Error('机筒模块不属于当前机型或未核对型号');
   const item=known.get(m.key),open=m.open??true;
   if(typeof m.uid!=='string'||!/^[A-Za-z0-9_-]{1,80}$/.test(m.uid)||ids.has(m.uid))throw Error('机筒模块标识无效或重复');
   if(typeof open!=='boolean')throw Error('排气口状态必须为布尔值');ids.add(m.uid);
   let cap=m.cap??(open?'open':'closed');
   if(item.caps.length){if(!item.caps.includes(cap))throw Error('堵头与当前机筒开口不兼容')}
   else{if((m.cap??'open')!=='open')throw Error('该模块没有可安装堵头的开口');cap='open'}
   const row={uid:m.uid,key:m.key,open:cap!=='closed',cap};
   if(Object.hasOwn(m,'length')){
    const length=m.length;if(!Number.isInteger(length))throw Error('节长必须为整数毫米');
    if(['heat','flange'].includes(item.role)){if(length!==item.length)throw Error('隔热片及法兰长度按本机标准保留')}
    else if(length<Math.max(10,item.caps.length?(machine==='50'?72:80):10)||length>spec.section_length*4)throw Error('节长超出设计范围或小于开口投影宽度');
    row.length=length;
   }
   return row;
  })};
  if(Object.hasOwn(layout,'screw_length')){const n=layout.screw_length;if(!Number.isInteger(n)||n<100||n>MAX_MODULES*spec.section_length*4+1000)throw Error('方案螺杆目标长度无效');result.screw_length=n}
  return result;
 }
 function rows(machine,spec,ports={}){
  let mm=0;const layout=ports.barrel_layout,items=layout==null?spec.barrel_configuration:normalize(machine,spec,layout).modules,known=new Map(catalog(spec).map(x=>[x.key,x]));
  return items.map((x,i)=>{
   let r;if(layout==null){r={...x,start:mm,length:x.mm-mm,machine,nominal_diameter:spec.diameter,open:ports[x.port_key]??true};r.cap=r.open?'open':'closed'}
   else{r={...known.get(x.key),...x,pos:position(i),start:mm};r.mm=mm+r.length;r.port_key=PORT_ROLES.includes(r.role)?'module-'+r.uid:''}
   mm=r.mm;r.annotation=r.cap==='inject'?'INJECT':r.cap==='closed'?'CLOSED':({feed:'FEED',natural:'ATM',side:'SIDE FEED',vacuum:'VAC',side_vacuum:'SIDE VAC'}[r.role]||(r.role==='heat'?`${r.length} mm`:''));
   r.accessory=r.cap==='closed'?(r.role==='side'?'侧喂料堵头':r.role==='side_vacuum'?'侧真空堵头':'顶部全封闭堵头'):r.cap==='inject'?'中心注液堵头':'无堵头';
   r.portKey=r.port_key||(PORT_ROLES.includes(r.role)?'standard-'+i:'');r.fixed=!r.portKey;return r;
  });
 }
 function materialize(machine,spec,ports){
  if(ports.barrel_layout!=null)return normalize(machine,spec,ports.barrel_layout);
  return {version:1,machine,modules:rows(machine,spec,ports).map((r,i)=>{const item=catalog(spec).find(c=>c.name===r.name&&c.role===r.role&&c.length===r.length);if(!item)throw Error('标准机筒模块尚未核对，无法编辑');return {uid:'standard-'+i,key:item.key,open:r.open,cap:r.cap}})};
 }
 function zones(machine,spec,ports){const width=machine==='50'?72:80;return rows(machine,spec,ports).flatMap((r,i)=>{if(!['natural','side','vacuum','side_vacuum'].includes(r.role)||!r.open)return [];const center=r.start+r.length/2;return [{section:i+1,kind:r.role,start:center-width/2-30,end:center+width/2+30}]})}
 function targetLength(machine,spec,ports){return ports.barrel_layout!=null?(normalize(machine,spec,ports.barrel_layout).screw_length??spec.element_length):spec.element_length}
 function summary(machine,spec,ports){const rs=rows(machine,spec,ports),process=rs.filter(r=>!['heat','flange'].includes(r.role)),active=process.reduce((n,r)=>n+r.length,0),length=rs.reduce((n,r)=>n+r.length,0);return {sections:process.length,active_length:active,process_d:Math.round(active/spec.section_length*4*1000)/1000,barrel_length:length,suggested_screw_length:spec.element_length+length-spec.barrel_length}}
 function warnings(machine,spec,ports){
  if(ports.barrel_layout==null)return [];const rs=rows(machine,spec,ports),length=rs.reduce((n,r)=>n+r.length,0),notes=['自定义机筒配置：需复核筒体内径、间隙、轴长及开口位置，未核对前不可直接用于生产'];
  if(!rs.length)notes.push('机筒配置为空');if(length!==spec.barrel_length)notes.push(`机筒总长 ${length} mm，与标准 ${spec.barrel_length} mm 不符；标准螺杆轴长不自动延长`);
  if(rs.filter(r=>r.role==='feed'&&r.open).length!==1)notes.push('主喂料模块数量不为1');if(rs.filter(r=>r.role==='flange').length!==1)notes.push('连接法兰数量不为1');
  if(targetLength(machine,spec,ports)!==spec.element_length)notes.push('已指定方案螺杆目标长度：实际轴长、强度及驱动适配需工程复核');return notes;
 }
 return {MAX_MODULES,PORT_ROLES,catalog,position,normalize,rows,materialize,zones,warnings,targetLength,summary};
});
